odl/solvers/util/steplen.py

```diff
@@ -118,16 +138,20 @@ def __call__(self, x, direction, dir_derivative):
             The current point
         direction : `LinearSpaceElement`
             Search direction in which the line search should be computed
-        dir_derivative : float
+        dir_derivative : float, optional
             Directional derivative along the ``direction``
+            Default: ``function.gradient(x).inner(direction)``
```

Comment:
Well, currently the code anyway assumes that `function` is a `Functional` due to [this line](https://github.com/odlgroup/odl/pull/758/files/26b163d8775a73c9227761fc1f983552692729d5#diff-4af2b36cbd22929eb22d8c1dd57afe8eR126). If this is fixed, then the user sees this when trying to call the line search instance:
![backtrack](https://cloud.githubusercontent.com/assets/5030250/21309562/1e0051da-c5df-11e6-9afc-0cffc800eb77.png)
If we focus on the interpreter & inspector part we see (already fading out) the help tooltip for the signature of the object:
![backtrack](https://cloud.githubusercontent.com/assets/5030250/21309684/8a725034-c5df-11e6-8c61-769e7b8120cf.png)
So `dir_derivative` is optional, right? No trace of the doc explaining *when* it is optional. And, of course, leaving it empty will fail:
```
s(0, 1)
Traceback (most recent call last):

  File "<ipython-input-4-3201524aa66e>", line 1, in <module>
    s(0, 1)

  File "/export/scratch1/kohr/git/odl/odl/solvers/util/steplen.py", line 153, in __call__
    dir_derivative = self.function.gradient(x).inner(direction)

AttributeError: 'function' object has no attribute 'gradient'
```
I don't think this is very user-friendly, even towards advanced users.

Of course, Spyder *should* display the doc when calling the object, so there's a bug, but IMO restrictions on the input parameters should *always* be documented in `__init__`, which also happens to be displayed more reliably.